src/v/raft/state_machine_manager.cc

```diff
@@ -334,26 +358,21 @@ ss::future<> state_machine_manager::apply() {
           _next, _raft->committed_offset(), ss::default_priority_class());
 
         model::record_batch_reader reader = co_await _raft->make_reader(config);
-        // collect STMs which has the same _next offset as the offset in
-        // manager and there is no background apply taking place
-        std::vector<entry_ptr> machines;
-        for (auto& [_, entry] : _machines) {
-            /**
-             * We can simply check if a mutex is ready here as calling
-             * maybe_start_background_apply() will make the mutex underlying
-             * semaphore immediately not ready as there are no scheduling points
-             * before calling `get_units`
-             */
-            if (
-              entry->stm->next() == _next
-              && entry->background_apply_mutex.ready()) {
-                machines.push_back(entry);
-            }
-        }
+
         auto last_applied = co_await std::move(reader).consume(
           batch_applicator(default_ctx, machines, _as, _log),
           model::no_timeout);
-
+        if (model::next_offset(last_applied) == _next) {
```

Comment:
IIUC `last_applied` can be `model::offset{}` if none of the state machines applied the batch successfully. Should `==` be `<=` here?

Reply:
I mean, in this consume attempt, with `_next` already having some value. Can't `last_applied` coming from `batch_applicator` be invalid here?